doc/developer/design/20220330_persist.md

```diff
@@ -0,0 +1,1139 @@
+- `seqno_since` frontier (the last `seqno` any reader will be allowed to read
+  from) I believe this makes the most sense as the min of (all of the readers
+  `seqno_since`, the `seqno` of the last state rollup?)
+
+Foreach active (non-expired) writer:
+  - the writer’s `upper` frontier
+  - the unique key prefix assigned to the writer
+  - the writer’s lease expiration time
+
+Foreach active (non-expired) reader:
+  - the reader’s `since` frontier
+  - the reader’s `seqno_since` frontier (the `seqno` before which it will never
+    read any metadata information)
+  - the reader’s lease expiration time
+
+Forall expired readers and writers:
+ - the expired reader / writer `ID`
+ - the `seqno` at which the reader/writer was expired
+
+For all submitted non deleted batches:
+ - the writer ID that submitted the batch
+ - the batch’s `[lower, upper)`, and `since` frontiers
+ - The key / set of keys pointing to the batch in S3
+  - (maybe?) the `seqno` when the batch was introduced
+
+Forall logically deleted batches:
+ - the writer that deleted the batch
+ - the `seqno` at which the batch was logically deleted
+ - the key / set of keys pointing to the batch in S3
+
+Foreach state rollup in the set state metadata rollups:
+ - the `seqno` associated with the state rollup
+- the key pointing to the state rollup in S3.
+
+Foreach logically deleted state rollup:
+ - the `seqno` at which the state rollup was deleted.
+
+This metadata state is stored in a sequence of incremental updates/individual
+state transitions in `Consensus`, and a set of state rollups stored in S3
+objects, for ease of access to historical data. The batches themselves are
+stored in S3 objects inside of writer-owned paths within `Blob`.
+
+#### Allowed state transitions:
+
+- `AddReaderWriter(lease_duration_seconds) -> Reader/Writer Id + WriteCapability
+  (upper) and ReadCapability(since, seqno_since)` - this state transition adds a
+  new reader and writer to the set of readers/writers associated with the same
+  id and the corresponding `since`/`seqno_since`/`upper` chosen from the global
+  values. The new readers and writers have a lease of `lease_duration_seconds`.
+
+- `LogicallyDeleteReader(reader_id)` / `LogicallyDeleteWriter(writer_id)`:
+  expire a reader / writer. All future requests by this reader or writer will be
+  rejected.
+
+- `PhysicallyDeleteReader(reader_id)` / `PhysicallyDeleteWriter(writer_id)`:
+  remove the reader / writer from the set of logically deleted readers and
+  writers.
+
+- `Append(batch_key, description, writer_id)`: this transition adds a batch to
+  the set of active batches and downgrades the writer’s `upper` to
+  `description.upper`. This transition can also downgrade the collection’s
+  `upper` if the global maximum of all writers’ `upper`s increases. This
+  transition also updates the writer’s lease expiration time.
+
+- `DowngradeReader(reader_id, new_since, new_seqno_since)`: Downgrade the reader
```

Comment:
`new_seqno_since` is hidden from the user, it's part of the state transition but maintained internally and doesn't show up on the public api of ReadHandle. `new_since` yeah has to be beyond the current `since`. in the way things are described here, that means the `since` of the handle. if we do the big downgrade_since change described in the notice of the doc, it would be the shard-level `since`